src/main.zig

```diff
@@ -385,6 +386,69 @@ const State = struct {
         try self.channel.flush();
     }
 
+    pub fn sendNotification(self: *State, method: []const u8, params: anytype) !void {
+        const params_bytes = try std.json.stringifyAlloc(self.allocator, params, .{});
+        defer self.allocator.free(params_bytes);
+
+        const notification = Notification{
+            .method = method,
+            .params = .{ .raw = params_bytes },
+        };
+
+        const format_options = std.json.StringifyOptions{
+            .emit_null_optional_fields = false,
+        };
+
+        var counting = std.io.countingWriter(std.io.null_writer);
+        try std.json.stringify(notification, format_options, counting.writer());
+        const content_length = counting.bytes_written;
+
+        const writer = self.channel.writer();
+        try writer.print("Content-Length: {}\r\n\r\n", .{content_length});
+        try std.json.stringify(notification, format_options, writer);
+        try self.channel.flush();
+    }
+
+    pub fn publishDiagnostics(self: *State, document: *Workspace.Document) !void {
+        var parse_diagnostics = std.ArrayList(parse.Diagnostic).init(self.allocator);
+        defer parse_diagnostics.deinit();
+
+        var tree = parse.parse(self.allocator, document.source(), .{
+            .diagnostics = &parse_diagnostics,
+        }) catch |err| {
+            std.log.warn("failed to re-parse for diagnostics: {}", .{err});
+            return;
+        };
+        defer tree.deinit(self.allocator);
```

Comment:
See if you can use the `Document.parseTree` function here instead. It caches the parsed tree, so that we don't have to redo the parsing next time the document is used (such as when triggering completions).